third_party/terraform/resources/resource_google_project_services.go

```diff
@@ -48,330 +54,321 @@ func resourceGoogleProjectServices() *schema.Resource {
+// services for a set of project services.
+func setServiceUsageProjectEnabledServices(services []string, project string, config *Config) error {
+	enabledSet, err := getEnabledServiceSet(project, config, ignoredProjectServicesSet)
+	if err != nil {
+		return err
+	}
+	servicesSet := golangSetFromStringSlice(services)
+	toEnable := make([]string, 0, len(servicesSet))
+	for srv := range servicesSet {
+		if _, ok := enabledSet[srv]; !ok {
+			toEnable = append(toEnable, srv)
 		}
-		return sm
 	}
 
-	sort.Strings(cfgServices)
-	cfgMap := m(cfgServices)
-	log.Printf("[DEBUG]: Saw the following services in config: %v", cfgServices)
-	apiMap := m(apiServices)
-	log.Printf("[DEBUG]: Saw the following services enabled: %v", apiServices)
-
-	for k := range apiMap {
-		if _, ok := cfgMap[k]; !ok {
-			log.Printf("[DEBUG]: Disabling %s as it's enabled upstream but not in config", k)
-			err := disableService(k, pid, config, true)
-			if err != nil {
-				return err
+	if err := enableServiceUsageProjectServices(services, project, config); err != nil {
+		return fmt.Errorf("unable to enable Project Services %s (%+v): %s", project, services, err)
+	}
+
+	for srv := range enabledSet {
+		if _, ok := servicesSet[srv]; !ok {
+			log.Printf("[DEBUG] Disabling project %s service %s", project, srv)
+			if err := disableServiceUsageProjectService(srv, project, config, true); err != nil {
+				return fmt.Errorf("unable to enable Project Services %s/%s): %s", project, srv, err)
 			}
-		} else {
-			log.Printf("[DEBUG]: Skipping %s as it's enabled in both config and upstream", k)
-			delete(cfgMap, k)
 		}
 	}
+	return nil
+}
 
-	keys := make([]string, 0, len(cfgMap))
-	for k := range cfgMap {
-		keys = append(keys, k)
-	}
-	sort.Strings(keys)
-	log.Printf("[DEBUG]: Enabling the following services: %v", keys)
-	err := enableServices(keys, pid, config)
+func disableServiceUsageProjectService(service, project string, config *Config, disableDependentServices bool) error {
+	err := retryTime(func() error {
+		name := fmt.Sprintf("projects/%s/services/%s", project, service)
+		sop, err := config.clientServiceUsage.Services.Disable(name, &serviceusage.DisableServiceRequest{
+			DisableDependentServices: disableDependentServices,
+		}).Do()
+		if err != nil {
+			return err
+		}
+		// Wait for the operation to complete
+		waitErr := serviceUsageOperationWait(config, sop, "api to disable")
+		if waitErr != nil {
+			return waitErr
+		}
+		return nil
+	}, 10)
 	if err != nil {
-		return err
+		return fmt.Errorf("Error disabling service %q for project %q: %v", service, project, err)
 	}
 	return nil
 }
 
-// Retrieve services defined in a config
-func getConfigServices(d *schema.ResourceData) (services []string) {
-	if v, ok := d.GetOk("services"); ok {
-		for _, svc := range v.(*schema.Set).List() {
-			services = append(services, svc.(string))
-		}
+func readEnabledServiceUsageProjectServices(project string, config *Config) ([]string, error) {
```

Comment:
Since this and `getEnabledServiceSet` are so similar, would we be able to drop one or the other and perform type conversions inline at the callsite, or give them the same arguments with different return types?

Reply:
Yeah, it's a bit silly right now, but my intention is that readEnabledServiceUsageProjectService will call out to the batcher, and the batcher will use getEnabledServiceSet as a callback